doc/shared/suncontrol/SUNControl_Description.rst

```diff
@@ -0,0 +1,480 @@
+   :param dsm:  the successful temporal error estimate.
+   :return: error code indicating success failure
+            (see :numref:`SUNControl.Description.errorCodes`).
+
+   Usage:
+
+   .. code-block:: c
+
+      retval = SUNControlUpdate(C, h, dsm);
+
+.. c:function:: int SUNControlUpdateMRIH(SUNControl C, realtype H, realtype h, realtype DSM, realtype dsm)
+
+   Notifies the controller of a successful multirate time step of sizes *H* and
+   *h*, and with temporal error estimates *DSM* and *dsm*. This is used for
+   controllers of type *SUNDIALS_CONTROL_MRI_H* that store a history of either
+   step size inputs or resulting error estimates for performing the estimation
+   process.
+
+   :param C:  the :c:type:`SUNControl` object..
+   :param H:  the successful slow step size.
+   :param h:  the successful fast step size.
+   :param DSM:  the successful slow temporal error estimate.
+   :param dsm:  the successful fast temporal error estimate.
+   :return: error code indicating success failure
+            (see :numref:`SUNControl.Description.errorCodes`).
+
+   Usage:
+
+   .. code-block:: c
+
+      retval = SUNControlUpdateMRIH(C, H, h, DSM, dsm);
+
+.. c:function:: int SUNControlUpdateMRITol(SUNControl C, realtype H, realtype tolfac, realtype DSM, realtype dsm)
+
+   Notifies the controller of a successful multirate time step of size *H* and
+   fast tolerance factor *tolfac*, that resulted in temporal error estimates
+   *DSM* and *dsm*. This is typically used for controllers of type
+   *SUNDIALS_CONTROL_MRI_TOL* that store a history of either control inputs or
+   resulting error estimates for performing the estimation process.
+
+   :param C:  the :c:type:`SUNControl` object..
+   :param H:  the successful slow step size.
+   :param tolfac:  the successful fast relative tolerance factor.
+   :param DSM:  the successful slow temporal error estimate.
+   :param dsm:  the successful fast temporal error estimate.
+   :return: error code indicating success failure
+            (see :numref:`SUNControl.Description.errorCodes`).
+
+   Usage:
+
+   .. code-block:: c
+
+      retval = SUNControlUpdate(C, h, dsm);
+
+.. c:function:: int SUNControlSpace(SUNControl C, long int *lenrw, long int *leniw)
+
+   Informative routine that returns the memory requirements of the
+   :c:type:`SUNControl` object.
+
+   :param C:  the :c:type:`SUNControl` object..
+   :param lenrw: (output)  number of ``sunrealtype`` words stored in the
+                 controller.
+   :param leniw: (output)  number of ``sunindextype`` words stored in the
+                 controller. This may also include pointers, `int` and
+                 `long int` words.
+   :return: error code indicating success failure
+            (see :numref:`SUNControl.Description.errorCodes`).
+
+   Usage:
+
+   .. code-block:: c
+
+      retval = SUNControlSpace(C, &lenrw, &leniw);
+
+
+
+.. _SUNControl.Description.errorCodes:
+
+SUNControl Error Codes
```

Comment:
I'll note that the `SUNAdaptController` and `SUNTimestepHeuristics` modules live outside of ARKODE, and are instead at the higher SUNDIALS-wide level (where we do not yet have shared error codes).  Are you recommending that instead of having error codes for these new modules, they instead use 0, >0, <0 to indicate success, recoverable failure, and nonrecoverable failure?

Reply:
These should have their own error codes since they live outside of ARKODE.

Reply:
I meant that `SUNAdaptController` and `SUNTimestepHeuristics` could return `>0`, `0`, or `<0` and ARKODE would internally map that to its own values similar to what we do for callbacks.

Reply:
I think error codes are certainly better than just using `>0`, `0` or `<0`, although we probably have some other places where we should have error codes but are just doing `>0`, `0` or `<0`.

Reply:
I think it's good for our implementations to use a consistent set of error codes but I'm not sure it's necessary for a user-supplied controller have to use these codes. The integratos are not worried about the specific error value just that it was `>0`, `0`, or `<0` e.g., for any `<0` return ARKODE could return `ARK_CONTROLLER_ERROR` and the user could query the controller for it's last error if necessary (like `GetLastFlag` with the linear solvers).

Reply:
We have defined error codes for all of our core classes though, so I think it is most consistent to have these error codes defined.

Reply:
I don't disagree that we should have defined error codes for these classes. I think we should clarify that since the integrators do not rely on the specific return value user-defined implementations of these classes are not required to use these error codes.